### PR TITLE
Add semi-structured sparsity to hf eval

### PR DESCRIPTION
**Pull request by jcaip**
Summary:

This PR adds in huggingface eval for semi-structured sparsity with the `nm-testing/SparseLlama-3-8B-pruned_50.2of4` checkpoint published by neuralmagic. We can accelerate this checkpoint with `to_sparse_semi_structured` with minimal accuracy loss. 

Also adds in an example notebook that outlines how to do text generation with huggingface's model and shows a 80 tok/s -> 87 tok/s speedup. 

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: scripts/hf_eval.py [hunk 1]
```diff
@@ -57,12 +62,29 @@ def run_evaluation(repo_id, tasks, limit, device, precision, quantization, compi
...
+        sparsify_(model, semi_sparse_weight(), filter_fn=all_linear)
+
     with torch.no_grad():
         result = evaluate(
             HFLM(
                 pretrained=model.to(device),
                 tokenizer=tokenizer,
-                batch_size=batch_size,
+                batch_size="auto",
+                max_batch_size=2048,
```
Comment (msaroufim) [inline scripts/hf_eval.py:87]: how come?

**Comment by jcaip** [comment]
mkn | dense_time | sparse_time | dense_memory | sparse_memory | speedup
-- | -- | -- | -- | -- | --
(1, 4096, 4096) | 0.026794 | 0.048279 | 0.050628 | 0.136668 | 0.553
(1, 4096, 1024) | 0.014931 | 0.048054 | 0.025443 | 0.046953 | 0.310
(1, 4096, 14336) | 0.098401 | 0.077046 | 0.134575 | 0.435717 | 1.277
(1, 14336, 4096) | 0.084088 | 0.153705 | 0.134534 | 0.435533 | 0.547
(8, 4096, 4096) | 0.028638 | 0.045108 | 0.050800 | 0.136783 | 0.635
(8, 4096, 1024) | 0.015131 | 0.044912 | 0.025529 | 0.047025 | 0.337
(8, 4096, 14336) | 0.099945 | 0.073439 | 0.135034 | 0.435975 | 1.364
(8, 14336, 4096) | 0.083806 | 0.148976 | 0.134850 | 0.435791 | 0.563
(16, 4096, 4096) | 0.029405 | 0.045662 | 0.050996 | 0.136979 | 0.642
(16, 4096, 1024) | 0.020930 | 0.045057 | 0.025628 | 0.047123 | 0.463
(16, 4096, 14336) | 0.100186 | 0.073811 | 0.135558 | 0.436499 | 1.365
(16, 14336, 4096) | 0.084441 | 0.150410 | 0.135210 | 0.436151 | 0.561

File: tutorials/huggingface_24sparse_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,383 @@
+{
```
Comment (msaroufim) [inline tutorials/huggingface_24sparse_example.ipynb:1]: Really not a fan of checking in notebooks, mind turning this into a python file, there should be a button to do this in jupyter, alternatively link to the tutorial hosted on google colab

Comment (jcaip) [reply tutorials/huggingface_24sparse_example.ipynb:1]: Sure, that fine with me.

File: tutorials/huggingface_24sparse_example.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+from tqdm import tqdm
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+os.environ["TOKENIZERS_PARALLELISM"] = "false" # silence warnings when compiling
+
+torch.sparse.SparseSemiStructuredTensor._FORCE_CUTLASS = True
+torch.set_float32_matmul_precision('high')
+
+def timed(fn):
```
Comment (msaroufim) [inline tutorials/huggingface_24sparse_example.py:19]: we already have timing utilities in torchao/utils.py

**Review by msaroufim — APPROVED** [review]
